job-runner/chain/traverser.go

```diff
@@ -81,209 +107,159 @@ func (f *traverserFactory) Make(jobChain proto.JobChain) (Traverser, error) {
-				t.retrySequence(doneJob) // re-enqueue first job of failed sequence
-				continue JOB_REAPER
-			}
-
-			// If the failed job is not part of a retryable sequence, then ignore
-			// subsequent jobs. For example, if job B of A -> B -> C  fails, then C is
-			// not ran.
-			jLogger.Warn("job did not complete successfully")
-			continue JOB_REAPER
-		}
-
-		// Job completed successfully, so enqueue/run the next jobs in the chain.
-		// This will yield multiple next jobs when the current job is the start of
-		// a sequence (a fanout node).
-		jLogger.Infof("job completed successfully")
-
-	NEXT_JOB:
-		for _, nextJ := range t.chain.NextJobs(doneJob.Id) {
-			nextJLogger := jLogger.WithFields(log.Fields{"next_job_id": nextJ.Id})
-
-			// Check to make sure the job is ready to run. It might not be if it has
-			// upstream dependencies that are still running.
-			if !t.chain.JobIsReady(nextJ.Id) {
-				nextJLogger.Infof("next job is not ready to run - not enqueuing it")
-				continue NEXT_JOB
-			}
-
-			nextJLogger.Infof("next job is ready to run - enqueuing it")
-
-			// Copy the jobData from the job that just finished to the next job.
-			// It is important to note that when a job has multiple parent
-			// jobs, it will get its jobData from whichever parent finishes
-			// last. Therefore, a job should never rely on jobData that was
-			// created during an unrelated sequence at any time earlier in
-			// the chain.
-			for k, v := range doneJob.Data {
-				nextJ.Data[k] = v
+	// Add the first jobs to the runJobChan.
+	// Prevent Stop()/shutdown() from closing runJobChan and don't enqueue jobs
+	// if the traverser has already been stopped or shutdown.
+	t.stopMux.Lock()
+	select {
+	case <-t.stopChan:
+		return
+	case <-t.shutdownChan:
+		return
+	default:
+		for _, job := range jobsToRun {
+			t.logger.Infof("sending initial job (%s) to runJobChan", job.Id)
+			if t.chain.IsSequenceStartJob(job.Id) {
+				// Starting a sequence, so increment sequence try count.
+				t.chain.IncrementSequenceTries(job.Id)
+				seqLogger := t.logger.WithFields(log.Fields{"sequence_id": job.SequenceId, "sequence_try": t.chain.SequenceTries(job.Id)})
+				seqLogger.Info("starting try of sequence")
 			}
-
-			t.runJobChan <- nextJ // add the job to the run queue
+			t.runJobChan <- job
 		}
+	}
+	t.stopMux.Unlock()
 
-		// Update chain repo for jobs next jobs just started ^
-		t.chainRepo.Set(t.chain)
+	// Reap jobs as they finish running. Stop reaping jobs when the doneJobChan
+	// is closed by runJobs - when the chain is done, stopped, or suspended.
+	for doneJob := range t.doneJobChan {
+		// Don't allow closing runJobChan / switching the reaper while we are
+		// reaping a job. Reap() should return quickly, so this is ok.
+		t.stopMux.Lock()
+		t.reaper.Reap(doneJob)
+		t.stopMux.Unlock()
 	}
 
-	return nil
+	// Indicate that the chain is done running. Stops waitForShutdown() listener.
+	select {
+	case <-t.doneChan:
+	default:
+		close(t.doneChan)
```

Comment:
I don't understand this. The only place `doneChan` is closed is here, but this also checks if `doneChan` is closed. The code comment doesn't clarify either. So I cannot tell what/how/when `doneChan` would be closed.